tools/autograd/templates/Functions.cpp

```diff
@@ -408,6 +408,27 @@ Tensor smooth_l1_loss_double_backward_grad_output(const Tensor & grad, const Ten
   return (r * grad).sum().view({1});
 }
 
+Tensor diag_backward(const Tensor & grad, const Tensor & self, int64_t diagonal) {
+  auto ndimension = self.ndimension();
+  TORCH_ASSERT(ndimension == 1 || ndimension == 2);
+
+  auto grad_input = grad.diag(diagonal);
+  if (ndimension == 1 || self.size(0) == self.size(1)) {
+    return grad_input;
+  }
+
+  // cat rows or cols to grad_input so that it matches self's shape.
+  auto length = grad_input.size(0);
+  auto self_nrows = self.size(0);
+  auto self_ncols = self.size(1);
+  if (self_nrows == length) {
+    auto extra_cols = grad_input.type().zeros({self_nrows, self_ncols - length});
+    return at::cat({grad_input, extra_cols}, 1);
+  }
+  auto extra_rows = grad_input.type().zeros({self_nrows - length, self_ncols});
+  return at::cat({grad_input, extra_rows});
```

Comment:
I think there's a faster way to fill a diagonal (`stride + 1` trick is the important part here):
```cpp
auto grad_input = self.type().zeros(self.sizes());
auto diag = grad_input.as_strided({diag_length}, {self.size(1) + 1});
diag.copy_(grad);
return grad_input;
```

Reply:
Thanks! Didn't realize this was a thing / this trick is pretty cool.